doc/developer/design/README.md

```diff
@@ -1,73 +1,98 @@
+# Design Process
 
-# What's the purpose of a design document?
+## What's the purpose of a design document?
 
-A design doc is a tool for the author to explain and gather feedback on a technical
-decision, providing an opportunity for anything they hadn't thought of to be
-surfaced early on and to have their thinking validated. It also serves as
+A design doc is a tool for the author to explain and gather feedback on a
+technical decision, providing an opportunity for anything they hadn't thought of
+to be surfaced early on and to have their thinking validated. It also serves as
 historical documentation for why a thing was done a certain way. The idea is to
-go slower at the beginning and get everything out on the table so we can go
-fast when implementing.
+go slower at the beginning and get everything out on the table so we can go fast
+when implementing.
 
-Design docs should contain enough information that a coworker could both justify and do the implementation rather than the author: this is a step along the path of transferring the work that makes it more likely "we", collectively, understand what we are looking at.
```

Comment:
I think we should still aim for that, and I think the existing README and the template say that. The _Detailed Description_ is still an important part of the doc.

The note for _Detailed Description_ is now

> Describe the approach in detail. If there is no clear frontrunner, feel free to
> list all approaches in alternatives. If applicable, be sure to call out any new
> testing/validation that will be required.
> 
> For some features it can be helpful to sketch an implementation. If you're
> working on things that are crossing team boundaries it will be helpful to spell
> out any new interfaces/traits/interactions.
> 
> For most new features, you should think about testing, rollout/lifecycle, and
> observability. These thinks can warrant their own sections.
 
Would you like to add some clarification to that/expand it? The motivation for removing that section is because it's very prescriptive in how detailed the doc should be. And I think we need to allow some leeway in interpreting that, based on the audience/the thing we're writing a doc for.

Reply:
> The motivation for removing that section is because it's very prescriptive in how detailed the doc should be. And I think we need to allow some leeway in interpreting that, based on the audience/the thing we're writing a doc for.

I would still keep the old text, maybe slightly relaxed and with the caveat that this is the end goal:

Design docs serve dual purpose:

1. During the actual design & implementation phase, the doc serves as a vehicle for driving discussion and establishing consensus around the available options and the selected alternative. During that phase the primary goal is to present alternatives and justify design decisions, and not so much on technical details.
2. After the implementation phase, the doc serves as a point-in-time artifact for an implemented design that helps disseminate knowledge. The primary goal for that phase is to record enough information so a coworker could review, diagnose, and debug the implementation of the design.

This means that in practice the "Detailed Description" can be refined to reflect things that were not known or spelled out during the initial design discussions, but turned out to be important and ultimately represent contextual information that will be important for somebody that needs to get familiar with the code in order to diagnose / debug / evolve the implementation.

Reply:
> Plus, we should expand our reference documentation, which is currently quite sparse (at least for the storage side).

So far I've not seen broad consensus and a specific policy on that front. Until we do this, I think we should restrain ourselves of relaxing the intentions of the design doc process too much. Alternatively, we can agree on something and just write it down in `doc/developer/design/README.md`. 

In general I think this

> consciously moving away from a fully described implementation towards a (hopefully) more streamlined writing/reviewing process, where we merge design docs earlier.

is _aboslutely fine_. My main concern is that _at some point during / after the implementation is done_ some kind of documents (design docs, reference docs, API docs, ideally consistent across all engineering teams) _should be_ updated in order to reflect an _up  to date_ description of the implemented design. This must not be too detailed, but sufficient so people can kick-start an exploration without having to schedule a dedicated intro session with the original author.

Reply:
> Alternatively, we can agree on something and just write it down in `doc/developer/design/README.md`.

Yes, let's do this. The current design doc process is not serving the broader org well; it's heavyweight enough that many engineers are not writing design documents at all, never mind keeping them up to date as the implementation progresses.

Reply:
I added some suggestions below. LMK what y'all think.